[CMake] Default to Debug builds

mnvr opened the pull request:
When CMake is invoked manually to configure a Swift build, it fails
because no default value is specified for the build type:

    CMake Error at cmake/modules/SwiftUtils.cmake:94 (message):
      Unknown build type:
    Call Stack (most recent call first):
      CMakeLists.txt:98 (is_build_type_optimized)

Instead we take a cue from LLVM, and default to "Debug" if the
CMAKE_BUILD_TYPE is unspecified.

The changes are essentially a copy of the relevant stanza from
'llvm/CMakeLists.txt'. There is one deviation: we use "NOT DEFINED" in
the conditional instead of relying on the empty string being false-y.

compnerd reviewed (commented):
I'm kinda torn on this.  Debug is the default behavior of CMake, and it makes sense.  The change is correct for that.  However, it also makes swiftc extremely painful to use for building things like the stdlib.

gottesmm commented:
@compnerd Elaborate?

mnvr commented:
@compnerd To clarify, my motivation behind this PR was to reduce the number of mandatory parameters to get a working CMake configuration, in order to make it less daunting for new people to get started.

The particular default used is not something I have enough experience with the codebase to decide on confidently. I used "Debug" simply because that is the one that LLVM had used. If you think that "Release" would be a better fit for Swift, I'll happily change that.


---

BTW: I've edited the commit. The `DEFINED` check is apparently not idiomatic in this case (Reference: [CMake FAQ Entry](https://cmake.org/Wiki/CMake_FAQ#How_can_I_change_the_default_build_mode_and_see_it_reflected_in_the_GUI.3F)), and so I have removed that. Further, I've modified the log messages to better match those used a few lines below, in the declaration of `SWIFT_STDLIB_BUILD_TYPE`.

mnvr committed:
[CMake] Default to Debug builds

When CMake is invoked manually to configure a Swift build, it fails
because no default value is specified for the build type:

    CMake Error at cmake/modules/SwiftUtils.cmake:94 (message):
      Unknown build type:
    Call Stack (most recent call first):
      CMakeLists.txt:98 (is_build_type_optimized)

Instead we take a cue from LLVM, and specify a default value for
CMAKE_BUILD_TYPE.  By providing a default, we reduce the number of
mandatory CMake configuration parameters, and make it easier for
people to get started.

The conditional was taken from llvm/CMakeLists.txt.  The check for
CMAKE_CONFIGURATION_TYPES is necessary to account for
multi-configuration generators [1].

The default was taken to be "Debug", simply because it was the one
which LLVM used.

[1]: https://cmake.org/pipermail/cmake/2006-February/008371.html

compnerd commented:
The debug build of `swiftc` is really slow at building the stdlib.  That said, I still end up using the debug builds ... Im just wondering if others would be comfortable with that as the default.  I suppose as long as people are using the `build-script` to build by default, this doesn't matter too much.

compnerd commented:
@swift-ci please test

compnerd commented:
Thinking more about this, most people are going to be building with `build-script`.  If they are building directly with CMake, I think it is safe to assume that they know how to properly invoke CMake for their needs.  LGTM
